source/module_esolver/esolver_fp.cpp

```diff
@@ -155,43 +155,25 @@ void ESolver_FP::after_scf(const int istep)
                     this->pw_rhod->real2recip(this->pelec->charge->rho_save[is], this->pelec->charge->rhog_save[is]);
                 }
                 std::string fn =PARAM.globalv.global_out_dir + "/SPIN" + std::to_string(is + 1) + "_CHG.cube";
-                ModuleIO::write_cube(
-#ifdef __MPI
-                    this->pw_big->bz,
-                    this->pw_big->nbz,
-                    this->pw_rhod->nplane,
-                    this->pw_rhod->startz_current,
-#endif
+                ModuleIO::write_cube(GlobalC::Pgrid,
                     data,
                     is,
                     PARAM.inp.nspin,
                     istep,
                     fn,
-                    this->pw_rhod->nx,
-                    this->pw_rhod->ny,
-                    this->pw_rhod->nz,
                     this->pelec->eferm.get_efval(is),
                     &(GlobalC::ucell),
                     PARAM.inp.out_chg[1],
                     1);
                 if (XC_Functional::get_func_type() == 3 || XC_Functional::get_func_type() == 5)
                 {
                     fn =PARAM.globalv.global_out_dir + "/SPIN" + std::to_string(is + 1) + "_TAU.cube";
-                    ModuleIO::write_cube(
-#ifdef __MPI
-                        this->pw_big->bz,
-                        this->pw_big->nbz,
-                        this->pw_rhod->nplane,
-                        this->pw_rhod->startz_current,
-#endif
+                    ModuleIO::write_cube(GlobalC::Pgrid,
                         this->pelec->charge->kin_r_save[is],
                         is,
                         PARAM.inp.nspin,
                         istep,
```

Comment:
if you plan to write these quantities in comment lines, you can build a string in advance (before calling the write_cube), then pass the string to the function write_cube.
What I mean is, the write_cube should be the dual of read_cube (as much as possible), and I have mentioned what will be a good param list of a function called read_cube.

Reply:
Good idea, I will try this.